app/ws-rewriter.js

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+var debug = require('debug')('dashboard-proxy:server');
+var error = require('debug')('dashboard-proxy:server:error');
+var EventEmitter = require('events').EventEmitter;
+var nbstore = require('./notebook-store');
+var Promise = require('es6-promise').Promise;
+var urljoin = require('url-join');
+var util = require('util');
+var WebSocketClient = require('websocket').client;
+var WebSocketConnection = require('websocket').connection;
+var WebSocketServer = require('websocket').server;
+
+/**
+ * Websocket proxy between client/browser and kernel gateway which rewrites execution requests to
+ * the gateway, replacing numeric index values with the associated code block.
+ *
+ * Emits the following events:
+ * 		'request': function(req: WebSocketRequest, conn: WebSocketConnection)
+ * 			Fired when the websocket server handles a request from the client/browser.
+ *
+ * @param  {Object} args
+ * @param  {Server} args.server - HTTP(S) server instance
+ * @param  {String} args.host - kernel gateway host/domain
+ * @param  {String} args.basePath - kernel gateway base URL
+ * @param  {Function} args.sessionToNbPath - callback to return a notebook path for the session ID
+ */
+function WsRewriter(args) {
+    debug('setting up WebSocket proxy');
+    EventEmitter.call(this);
+
+    this._host = args.host;
+    this._basePath = args.basePath;
+    this._sessionToNbPath = args.sessionToNbPath;
+
+    // create a websocket server which will listen for requests coming from the client/browser
+    var wsserver = new WebSocketServer({
+        httpServer: args.server,
+        autoAcceptConnections: false,
+        maxReceivedFrameSize: Number.MAX_SAFE_INTEGER
+    });
+
+    wsserver.on('request', this._handleWsRequest.bind(this));
+}
+util.inherits(WsRewriter, EventEmitter);
+
+WsRewriter.prototype._handleWsRequest = function(req) {
+    debug('ws-server connection request');
+
+    var self = this;
+    var servConn = req.accept(null, req.origin); // accept from all origins
```

Comment:
Nothing to solve here. We'd solve it at the auth level. Something akin to https://github.com/expressjs/csurf where we render a token onto the HTML page of the dashboard and expect to see that token back in the request.

We might already be getting something like XSRF protection from the way we're using session ID. Another security review is probably in order. For now, leave it as-is.